include/nbl/builtin/hlsl/cpp_compat.hlsl

```diff
@@ -38,6 +45,60 @@ namespace nbl
 namespace hlsl
 {
 
+namespace impl
+{
+    template<class T> float16_t asfloat16_t(T val) { return asfloat16(val); }
+    template<class T> float32_t asfloat32_t(T val) { return asfloat(val); }
+    template<class T> int16_t   asint16_t(T val) { return asint16(val); }
+    template<class T> int32_t   asint32_t(T val) { return asint(val); }
+    template<class T> uint16_t  asuint16_t(T val) { return asuint16(val); }
+    template<class T> uint32_t  asuint32_t(T val) { return asuint(val); }
+    
+    template<class T>
+    float64_t asfloat64_t(T val) 
+    { 
+        uint64_t us = uint64_t(val);
+        return asdouble(uint32_t(val & ~0u), uint32_t((val >> 32u) & ~0u)); 
+    }
+    
+    template<>
+    float64_t asfloat64_t<float64_t>(float64_t val) { return val; }
+
+    template<class T> uint64_t asuint64_t(T val) {  return val; }
+    template<class T> int64_t asint64_t(T val) {  return asuint64_t(val); }
+    
+    template<>
+    uint64_t asuint64_t<float64_t>(float64_t val) 
+    { 
+        uint32_t lo, hi;
+        asuint(val, lo, hi);
+        return (uint64_t(hi) << 32u) | uint64_t(lo);
+    }
```

Comment:
this is really disgusting, it seems that HLSL is a bit dumb when it comes to converting doubles.

I'd take a look at #519 and the `spirv` header (copy it over), then make intrinsics for `OpConvertFtoU`, etc.
https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_conversion_instructions

also use our type-traits to constrain the `T` a little bit